rclcpp_lifecycle/src/lifecycle_node_state_manager.cpp

```diff
@@ -0,0 +1,432 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "lifecycle_node_state_services_manager.hpp"
+#include "lifecycle_node_state_manager.hpp"
+#include <cassert>
+#include <memory>
+
+namespace rclcpp_lifecycle
+{
+void
+LifecycleNodeStateManager::init(
+  const std::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface> node_base_interface,
+  const std::shared_ptr<rclcpp::node_interfaces::NodeServicesInterface> node_services_interface,
+  bool enable_communication_interface)
+{
+  using ChangeStateSrv = lifecycle_msgs::srv::ChangeState;
+  using GetStateSrv = lifecycle_msgs::srv::GetState;
+  using GetAvailableStatesSrv = lifecycle_msgs::srv::GetAvailableStates;
+  using GetAvailableTransitionsSrv = lifecycle_msgs::srv::GetAvailableTransitions;
+
+  node_base_interface_ = node_base_interface;
+
+  rcl_node_t * node_handle = node_base_interface_->get_rcl_node_handle();
+  const rcl_node_options_t * node_options =
+    rcl_node_get_options(node_base_interface_->get_rcl_node_handle());
+  auto state_machine_options = rcl_lifecycle_get_default_state_machine_options();
+  state_machine_options.enable_com_interface = enable_communication_interface;
+  state_machine_options.allocator = node_options->allocator;
+
+  // The call to initialize the state machine takes
+  // currently five different typesupports for all publishers/services
+  // created within the RCL_LIFECYCLE structure.
+  // The publisher takes a C-Typesupport since the publishing (i.e. creating
+  // the message) is done fully in RCL.
+  // Services are handled in C++, so that it needs a C++ typesupport structure.
+  std::lock_guard<std::recursive_mutex> lock(state_machine_mutex_);
+  state_machine_ = rcl_lifecycle_get_zero_initialized_state_machine();
+  rcl_ret_t ret = rcl_lifecycle_state_machine_init(
+    &state_machine_,
+    node_handle,
+    ROSIDL_GET_MSG_TYPE_SUPPORT(lifecycle_msgs, msg, TransitionEvent),
+    rosidl_typesupport_cpp::get_service_type_support_handle<ChangeStateSrv>(),
+    rosidl_typesupport_cpp::get_service_type_support_handle<GetStateSrv>(),
+    rosidl_typesupport_cpp::get_service_type_support_handle<GetAvailableStatesSrv>(),
+    rosidl_typesupport_cpp::get_service_type_support_handle<GetAvailableTransitionsSrv>(),
+    rosidl_typesupport_cpp::get_service_type_support_handle<GetAvailableTransitionsSrv>(),
+    &state_machine_options);
+  if (ret != RCL_RET_OK) {
+    throw std::runtime_error(
+            std::string("Couldn't initialize state machine for node ") +
+            node_base_interface_->get_name());
+  }
+
+  if (enable_communication_interface) {
+    state_services_manager_hdl_ = std::make_unique<LifecycleNodeStateServicesManager>(
```

Comment:
I'm a bit concerned about this interdependency of the the services manager and the state manager. Can this dependency be just one way (e.g. have the services manager depend on state manager but not vice versa)? This `enable_communication_interface` check could then be in the above `lifecycle_node_interface_impl` and ignore creating the services if disabled.

Reply:
I'm not a huge fan either. One difficulty is exposing the `state_machine_` to `StateServicesManager` for the construction of the `StateServicesManager` as the `state_machine_mutex_` would possibly be needed, going to re-diagram this out